File: src/chacana/ast.py
"""AST node types for Chacana expressions (MathJSON ValidationToken format)."""

from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class Variance(Enum):
    CONTRA = "Contra"
    COVAR = "Covar"


class IndexType(Enum):
    LATIN = "Latin"
    GREEK = "Greek"
    SPINOR = "Spinor"


# Canonical head strings — use these instead of raw string literals.
HEAD_ADD = "Add"
HEAD_NEGATE = "Negate"
HEAD_MULTIPLY = "Multiply"
HEAD_WEDGE = "Wedge"
HEAD_NUMBER = "Number"
HEAD_EXTERIOR_DERIVATIVE = "ExteriorDerivative"
HEAD_LIE_DERIVATIVE = "LieDerivative"
HEAD_TRACE = "Trace"
HEAD_DETERMINANT = "Determinant"
HEAD_INVERSE = "Inverse"
HEAD_HODGE_STAR = "HodgeStar"
HEAD_INTERIOR_PRODUCT = "InteriorProduct"
HEAD_PERTURBATION = "Perturbation"
HEAD_COMMUTATOR = "Commutator"
HEAD_DIVIDE = "Divide"
HEAD_RATIONAL = "Rational"


@dataclass(frozen=True)
class ChacanaIndex:
    label: str
    variance: Variance
    index_type: IndexType = IndexType.LATIN
    is_derivative: bool = False
    derivative_type: str | None = None  # "Semicolon" or "Comma"

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "label": self.label,
            "variance": self.variance.value,
            "type": self.index_type.value,
        }
        if self.is_derivative:
            result["is_derivative"] = True
            result["derivative_type"] = self.derivative_type
        return result


@dataclass
class TokenMetadata:
    symmetrized_groups: list[list[int]] = field(default_factory=list)
    antisymmetrized_groups: list[list[int]] = field(default_factory=list)
    order: int | None = None


@dataclass
class ValidationToken:
    head: str
    indices: list[ChacanaIndex] = field(default_factory=list)
    args: list[ValidationToken] = field(default_factory=list)
    value: float | None = None
    metadata: TokenMetadata = field(default_factory=TokenMetadata)

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {"type": "TensorExpression", "head": self.head}
        if self.indices:
            result["indices"] = [idx.to_dict() for idx in self.indices]
        if self.args:
            result["args"] = [arg.to_dict() for arg in self.args]
        if self.value is not None:
            result["value"] = self.value
        meta: dict[str, Any] = {}
        if self.metadata.symmetrized_groups:
            meta["symmetrized_groups"] = self.metadata.symmetrized_groups
        if self.metadata.antisymmetrized_groups:
            meta["antisymmetrized_groups"] = self.metadata.antisymmetrized_groups
        if self.metadata.order is not None:
            meta["order"] = self.metadata.order
        if meta:
            result["metadata"] = meta
        return result


def walk_tokens(token: ValidationToken) -> Iterator[ValidationToken]:
    """Pre-order traversal of a token tree."""
    yield token
    for arg in token.args:
        yield from walk_tokens(arg)

File: vscode-chacana/src/server/ast.ts
/**
 * Core AST types for Chacana tensor expressions.
 * Ports src/chacana/ast.py to TypeScript.
 */

export enum Variance {
  Contra = "Contra",
  Covar = "Covar",
}

export enum IndexType {
  Latin = "Latin",
  Greek = "Greek",
  Spinor = "Spinor",
}

export interface ChacanaIndex {
  label: string;
  variance: Variance;
  indexType: IndexType;
  isDerivative: boolean;
  derivativeType: "semicolon" | "comma" | null;
}

export interface SourceRange {
  startLine: number;
  startColumn: number;
  endLine: number;
  endColumn: number;
}

export interface TokenMetadata {
  symmetrized_groups: number[][];
  antisymmetrized_groups: number[][];
  order: number | null;
}

export function emptyMetadata(): TokenMetadata {
  return { symmetrized_groups: [], antisymmetrized_groups: [], order: null };
}

export interface ValidationToken {
  head: string;
  indices: ChacanaIndex[];
  args: ValidationToken[];
  value: number | null;
  metadata: TokenMetadata;
  range: SourceRange | null;
}

// Canonical head names (same as ast.py)
export const HEAD_ADD = "Add";
export const HEAD_NEGATE = "Negate";
export const HEAD_MULTIPLY = "Multiply";
export const HEAD_WEDGE = "Wedge";
export const HEAD_EXTERIOR_DERIVATIVE = "ExteriorDerivative";
export const HEAD_LIE_DERIVATIVE = "LieDerivative";
export const HEAD_TRACE = "Trace";
export const HEAD_DETERMINANT = "Determinant";
export const HEAD_INVERSE = "Inverse";
export const HEAD_HODGE_STAR = "HodgeStar";
export const HEAD_INTERIOR_PRODUCT = "InteriorProduct";
export const HEAD_PERTURBATION = "Perturbation";
export const HEAD_COMMUTATOR = "Commutator";
export const HEAD_NUMBER = "Number";

/** Map from functional operator identifier to canonical head. */
export const FUNCTIONAL_OP_MAP: Record<string, string> = {
  d: HEAD_EXTERIOR_DERIVATIVE,
  L: HEAD_LIE_DERIVATIVE,
  Tr: HEAD_TRACE,
  det: HEAD_DETERMINANT,
  inv: HEAD_INVERSE,
  star: HEAD_HODGE_STAR,
  hodge: HEAD_HODGE_STAR,
  i: HEAD_INTERIOR_PRODUCT,
};

/** Structural heads that are operators, not tensor names. */
export const STRUCTURAL_HEADS = new Set([
  HEAD_ADD,
  HEAD_NEGATE,
  HEAD_MULTIPLY,
  HEAD_WEDGE,
  HEAD_EXTERIOR_DERIVATIVE,
  HEAD_LIE_DERIVATIVE,
  HEAD_TRACE,
  HEAD_DETERMINANT,
  HEAD_INVERSE,
  HEAD_HODGE_STAR,
  HEAD_INTERIOR_PRODUCT,
  HEAD_PERTURBATION,
  HEAD_COMMUTATOR,
  HEAD_NUMBER,
]);

/** Greek lowercase range for index type detection. */
const GREEK_RE = /[\u03B1-\u03C9\u0391-\u03A9]/;

export function detectIndexType(label: string): IndexType {
  return GREEK_RE.test(label) ? IndexType.Greek : IndexType.Latin;
}

/** Pre-order traversal of a token tree. */
export function* walkTokens(token: ValidationToken): Generator<ValidationToken> {
  yield token;
  for (const arg of token.args) {
    yield* walkTokens(arg);
  }
}

export function makeToken(
  head: string,
  args: ValidationToken[] = [],
  indices: ChacanaIndex[] = [],
  value: number | null = null,
  metadata: TokenMetadata = emptyMetadata(),
  range: SourceRange | null = null,
): ValidationToken {
  return { head, args, indices, value, metadata, range };
}
